### ISSUE TITLE
Improve data path

### ISSUE THREAD

**Issue by zhanghongzhou** [sheepdog/sheepdog#156]
Hi experts,

I find that data path of IO request is complicated. Take this scenario as an example

1. Export sbd or sheepfs on client C1 is exported from server S1
2. Request from C1 for one data block whose 3-replica are stored on node S2, S3 and S4.
    C1->(req to)gateway service on S1(hash here)->(req to)peer service on S2, S3 or S4->(data to)S1-(data to)C1

This brings huge latency and network burden.

In Ceph it's much simpler:
C1(hash here)->(req to)S2, S3 or S3->(data to)C1

Can we simplify the whole process by
1. send from S2, S3 or S4 directly to C1
2. Add hash table into sbd or sheepfs

**Comment by mitake** [comment]
I didn't know the data path in ceph, it seems to be much simpler. However, according to our benchmarking result, sheepdog provides better performance than ceph. You can find results from P.22 - of this slide: http://sheepdog.github.io/sheepdog/_static/sheepdog_COJ14.pdf

> Add hash table into sbd or sheepfs

This is quite difficult because the hash table is updated by atomic broadcast of cluster managers (corosync and zookeeper). If we really want to work on this optimization, sheepdog clients (qemu, tgt, sbd, sheepfs, etc) must be linked with corosync and zookeeper's libraries. And the cost of changing will be quite high.

BTW, are you facing performance problem of sheepdog already?

**Comment by mitake** [comment]
If you want to optimize data path, LTTng infrastructure would be useful. @cjd3000 is working on analysis with it. I think you guys can cooperate on this topic :)

**Comment by zhanghongzhou** [comment]
Yes We ran some FIO tests from sbd, to compare with latest Ceph(hammer release 2015.6.9) rbd. Result is not good.

I saw in the COJ reports sheepdog beats Ceph in Qemu, iSCISI, KVM, pgbench tests. Maybe it's because Ceph improves a lot ever since? Or sbd itself is not optimized?

**Comment by mitake** [comment]
AFAIK, sbd is not a mature subsystem. Could you try QEMU and iSCSI?

**Comment by zhanghongzhou** [comment]
We're interested in sheepdog as block device. Can you shed some light on how we can improve sbd?To my understanding block device driver is simple. The gap might be in the sheepdog stack.
Just a guess.

**Comment by mitake** [comment]
iSCSI can be used as a block device (e.g. /dev/sda), of course. In addition, the block devices provided by iSCSI can be used by other systems than linux (e.g. Windows, VMware vSphere, etc).

If you really want to sbd, I'll dig in the performance problem. However, making sheepdog clients require amount of engineering costs. I suggest iSCSI if there's no special reason.

**Comment by zhanghongzhou** [comment]
Great suggestion. I'll try iSCSI and feedback the result to you.
Thanks.

**Comment by zhanghongzhou** [comment]
@mitake I tried iSCSI. IOPS is 20%~36% better than sbd(varies in test cases). 
Can we assume bottleneck is not in front-end? 
I am not familiar with the code stack, so is there a risk if we shorten the path by sending blocks from target node to front-end directly?

**Comment by mitake** [comment]
Thanks for sharing the result @zhanghongzhou . Could you share the detailed configuration, workload and performance?

I think current main bottlenecks are in thread model and grain of locking in sheep code. But I'm not sure which part is critical. So detailed analysis is required.
